packages/lyra/src/prefix-tree/trie.ts

```diff
@@ -107,14 +107,14 @@ export class Trie {
     return output;
   }
 
-  removeDocByWord(word: string, docID: string): boolean {
+  removeDocByWord(word: string, docID: string, exact = false): boolean {
     const root = this.root;
     if (!word) return false;
 
     function removeWord(node: TrieNode, _word: string, docID: string): boolean {
       const [nodeWord /**_docs*/] = node.getWord();
 
-      if (node.end && nodeWord === word) {
+      if (node.end || (exact && node.end && nodeWord === word)) {
```

Comment:
I don't think this fixes #40 . At a glance, #40 is caused by setting `end` to `false` without checking if the term is associated to other documents.

Reply:
I don't think so, #40 was caused because the document was deleted only when the words were the same (implementing `exact` parameter we don't have that issue). 

I think #38 is caused for the reason you said. I'm working on it.

Reply:
#38 cannot be caused by this, because it occurs even if no document is deleted, so this code path is never called in the reproduction.